sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,275 @@
+#pragma once
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{   
+
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter>&& exporter,
+                                       const int max_queue_size,
+                                       const int schedule_delay_millis,
+                                       const int max_export_batch_size)
+       :exporter_(std::move(exporter)),
+        max_queue_size_(max_queue_size), 
+        schedule_delay_millis_(schedule_delay_millis),
+        max_export_batch_size_(max_export_batch_size)   
+{  
+    buffer_ = std::unique_ptr<CircularBuffer<Recordable>>(
+                    new CircularBuffer<Recordable>(max_queue_size_));
+    
+    //Start the background worker thread
+    worker_thread_ = std::unique_ptr<std::thread>(
+                        new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+} 
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept 
+{
+    return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept  
+{
+    // no-op
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept 
+{
+    if(is_shutdown_ == true){
+        // TODO: log in glog
+        return;
+    }
+
+    std::unique_lock<std::mutex> lk(cv_m_);
+
+    if(static_cast<int>(buffer_->size()) >= max_queue_size_){
+        // TODO: glog that spans will likely be dropped
+    }
+
+    buffer_->Add(span);
+
+    // If the queue gets at least half full a preemptive notification is 
+    // sent to the worker thread to start a new export cycle.
+    if(static_cast<int>(buffer_->size()) >= max_queue_size_ / 2){
```

Comment:
Since this wasn't stated anywhere explicitly in the spec, I was wondering if it was fine to leave this as a hit-or-miss sort of a notify call. All it does is try to make a preemptive notification for better performance and is not a hard requirement.

But I think when there are many producer threads, this sort of preemption might become necessary to minimize the number of spans dropped.

Please let me know what you think about this!

Reply:
@reyang @pyohannes 
It'd be great to get your feedback too on this!

Reply:
Ok, if it is not a hard requirement, I think it is fine to leave it as a single attempt to wake up.

Reply:
It seems to be a good idea to me.